### ISSUE TITLE
EEG data classification

### ISSUE THREAD

**Issue by tressoldi** [mlr-org/mlr#1514]
I have different EEG data sets related to four conditions: no stimulation, 10,12 and 14 Hz stimulation. These three conditions are always 32, mixed within approximately 750 no stimulation. 
I simply need to know the precision of the classification of my learner for each data set independently.

I used the following syntax:
classif.task = makeClassifTask(data = data, target = "Stimulation")
lrn = makeFilterWrapper(learner="classif.naiveBayes", fw.method= "information.gain",
                        predict.type = "prob", fw.perc = 0.5, 
                        fix.factors.prediction = TRUE)

mod = train(lrn, classif.task)

rdesc = makeResampleDesc(method = "Subsample", stratify = TRUE)
r = resample(learner = lrn, task = classif.task, resampling = rdesc, show.info = FALSE)

pred = predict(mod, task = classif.task)
pred

print(calculateConfusionMatrix(pred, relative = T))

I'm missing something or I'm correct?

**Comment by larskotthoff** [comment]
What exactly are you predicting? In general, your code looks ok.

**Comment by berndbischl** [comment]
what really does not look OK is this:
you are cross-validating your learner first, that looks ok.
then you fit your learner on the complete data and evaluate it there, too. 
that is in-sample evaluation and usually not.

if you dont explicitly want that and know what you are doing, this looks very fishy.

**Comment by tressoldi** [comment]
I'm predicting the classification of my four "Stimulation" values: 0, 10, 12 and 14.
As I wrote. whereas the data for the 10,12 and 14 values are fixed (n=32), those related to 0 are approximately 750.

**Comment by berndbischl** [comment]
can you please comment on "in-sample evaluation" stuff.

and what feedback you explicitly would like to have from us?

**Comment by tressoldi** [comment]
I'd like to have your opinion if my "in-sample(s) evaluation" syntax may generate unreliable predictions. If yes, how I should change my syntax.

**Comment by berndbischl** [comment]
> I'd like to have your opinion if my "in-sample(s) evaluation" syntax may generate unreliable predictions. If yes, how I should change my syntax.

this is really not meant to be rude, but: 
have you read some kind of a book or other formal material for ML, and ML evaluation techniques?
because if not, we can hardly help you in the issue tracker here, except for guiding you towards appropriate material - that you then have to read yourself.

this does not subsitute a book or paper, but e.g. read here:
https://en.wikipedia.org/wiki/Overfitting

if you fit your model on data D, then evaluate on D as well, performance metrics will look potentially too optimistic. that what we call a biased evaluation, because it is "in-sample" = on the training set.

which happens in your code here:

mod = train(lrn, classif.task)
pred = predict(mod, task = classif.task)
print(calculateConfusionMatrix(pred, relative = T))

this has, at least as far as i understand your question, nothing todo with mlr syntax, but you needing to understand how ML evaluation works.  and if you don't understand your problem here, our best and honest advice to you must be: please read a book on the topic.

if i am simply misunderstanding your question please rephrase.

**Comment by tressoldi** [comment]
Probably I misunderstood the "train" function as explained in 
https://mlr-org.github.io/mlr-tutorial/devel/html/train/index.html

My interpretation is that it takes part of the data that can also specified for example with the 
train.set = sample(n, size = n/3) and, after the resampling, there is a prediction on the remaining data. 
What I misunderstood?

**Comment by schiffner** [comment]
Hi,

> My interpretation is that it takes part of the data that can also specified for example with the
train.set = sample(n, size = n/3) and, after the resampling, there is a prediction on the remaining data.
What I misunderstood?

By default, i.e., if you don't tell the `train` function that you want to train on a subset of the data, it will use **all available data**. The same applies for the `predict` function.
So what happens in your code is:
```
mod = train(lrn, classif.task)                                    ## trains the learner on the entire data set
pred = predict(mod, task = classif.task)                  ## makes predictions for the entire data set
print(calculateConfusionMatrix(pred, relative = T))  ## in-sample evaluation
```

This confusion matrix is not a reliable indicator of the generalization ability of your learner (i.e., the ability to correctly predict the class label of formerly unseen objects).

In order to assess this, you need to calculate the performance / confusion matrix based on the resampled predictions. You can do this as follows:
```
r = resample("classif.lda", sonar.task, resampling = makeResampleDesc("CV", iters = 10))
[Resample] cross-validation iter 1: mmce.test.mean=0.286
[Resample] cross-validation iter 2: mmce.test.mean=0.286
[Resample] cross-validation iter 3: mmce.test.mean=0.143
[Resample] cross-validation iter 4: mmce.test.mean=0.286
[Resample] cross-validation iter 5: mmce.test.mean=0.286
[Resample] cross-validation iter 6: mmce.test.mean=0.286
[Resample] cross-validation iter 7: mmce.test.mean= 0.3
[Resample] cross-validation iter 8: mmce.test.mean=0.35
[Resample] cross-validation iter 9: mmce.test.mean=0.333
[Resample] cross-validation iter 10: mmce.test.mean=0.19
[Resample] Aggr. Result: mmce.test.mean=0.275

## extract predictions
rpreds = getRRPredictions(r)

## confusion matrix
calculateConfusionMatrix(rpreds, relative = TRUE)
Relative confusion matrix (normalized by row/column):
        predicted
true     M         R         -err.-   
  M      0.73/0.75 0.27/0.30 0.27     
  R      0.28/0.25 0.72/0.70 0.28     
  -err.-      0.25      0.30 0.55     


Absolute confusion matrix:
        predicted
true      M  R -err.-
  M      81 30     30
  R      27 70     27
  -err.- 27 30     57
```

Below is for comparison the confusion matrix resulting from training and predicting on the same data set. The difference in the number of errors (57 vs. 20) is pretty impressive.

```
mod = train("classif.lda", sonar.task)
pred = predict(mod, sonar.task)
calculateConfusionMatrix(pred, relative = TRUE)

Relative confusion matrix (normalized by row/column):
        predicted
true     M         R         -err.-   
  M      0.93/0.90 0.07/0.09 0.07     
  R      0.12/0.10 0.88/0.91 0.12     
  -err.-      0.10      0.09 0.19     


Absolute confusion matrix:
        predicted
true       M  R -err.-
  M      103  8      8
  R       12 85     12
  -err.-  12  8     20

performance(pred)
      mmce 
0.09615385 
```

**Comment by tressoldi** [comment]
Thank you very much Julia for your clear explanation.

What are the differences if I use:
>train.set = sample(n, size = n/3)
>mod = train(lrn, classif.task)

and then
## extract predictions
>rpreds = getRRPredictions(mod)

## confusion matrix
>calculateConfusionMatrix(rpreds, relative = TRUE)
